p4-fusion/thread_pool.cc

```diff
@@ -76,25 +73,27 @@ void ThreadPool::ShutDown()
 		m_ThreadExceptions.clear();
 	}
 
-	SUCCESS("Thread pool shut down successfully");
+	SUCCESS("Thread pool shut down successfully")
 }
 
-ThreadPool::ThreadPool(int size)
+ThreadPool::ThreadPool(const int size, const std::string& repoPath, const bool fsyncEnable, const int tz)
+    : m_HasShutDownBeenCalled(false)
+    , m_ShouldStop(false)
 {
-	m_HasShutDownBeenCalled = false;
-	m_ShouldStop = false;
-
-	// Initialize the thread handlers;
-	m_Threads.resize(size);
-
+	// Initialize size threads.
 	for (int i = 0; i < size; i++)
 	{
-		Thread& t = m_Threads[i];
-		t.m_T = std::thread([this, &t, i]()
+		m_Threads.emplace_back([this, i, &repoPath, &fsyncEnable, &tz]()
 		    {
 				// Add some human-readable info to the tracing.
 				MTR_META_THREAD_NAME(("Worker #" + std::to_string(i)).c_str());
 
+			    // Initialize p4 API.
+			    P4API p4;
+			    GitAPI git(repoPath, fsyncEnable, tz);
+
```

Comment:
to avoid lock contention, the git api is now per-thread.

Reply:
It's non-obvious that thread it is safe to open the same repository and create objects in that repository from different threads concurrently.

I suggest at least adding a link to the discussion in https://github.com/libgit2/libgit2/issues/2491 which has these comments

> If you want multiple threads to share a single git_repository then you have to be more careful about which operations you use. We have made a lot of progress in thread safety, but I can't say with confidence that we have eliminated all of the race conditions. Simple object lookup and creation is fairly well tested

> One last situation is where you "solve" the one-thread-to-one-repository suggestion by having multiple git_repository objects that all refer to the same on-disk repo, one per thread. This is risky because cache validation is hard. Libgit2 tries hard to check cached values but since it only relies on timestamps and file sizes in some cases to decide if a file needs to be reloaded, some operations can run into caching issues. I would not think you'd see crashes in this case, but also some operations might run with stale data.